src/validator.js

```diff
@@ -171,15 +166,62 @@ export default class Validator {
     });
   }
 
+  scanFiles(files) {
+    return new Promise((resolve, reject) => {
+      // Resolve once every file in the XPI has been checked.
+      var promises = [];
+      for (let filename of files) {
+        promises.push(this.scanFile(filename));
+      }
+      return Promise.all(promises)
+        .then(() => {
+          resolve();
+        }).catch(reject);
+    });
+  }
+
+  getScanner(filename) {
+    switch (extname(filename)) {
+      case '.js':
+        return JavaScriptScanner;
+      default:
+        throw new Error('No scanner available for ${filename}');
+    }
+  }
+
+  scanFile(filename) {
+    return new Promise((resolve, reject) => {
+      // We might have to refactor this if we need to
+      // deal with streams *and* strings. But let's wait until that
+      // point comes. So far it seems most valiators want strings.
+      this.xpi.getFileAsString(filename)
+        .then((code) => {
+          let scanner = new (this.getScanner(filename))(code, filename);
+          return scanner.scan();
+        })
+        // messages should be a list of raw message data objects.
+        .then((messages) => {
+          for (let message of messages) {
+            if (typeof message.type === 'undefined') {
+              throw new Error('message.type must be defined');
+            }
+            this.collector._addMessage(message.type, message);
+          }
+          resolve();
+        })
+        .catch(reject);
+    });
+  }
+
   scan(_Xpi=Xpi) {
     return new Promise((resolve, reject) => {
       this.checkFileExists(this.packagePath)
         .then(() => {
           this.xpi = new _Xpi(this.packagePath);
-          return this.xpi.getJSFiles();
+          return this.xpi.getFilesByExt('.js');
```

Comment:
Yeah good point - there's quite a number of initial assumptions here - but that's fine we can refactor as we go.